Add LinkAttrs constructor NewLinkAttrs

LK4D4 opened the pull request:
It looks little awkward, but should be enough for people who care about
defaults from kernel.

Fix #26

LK4D4 commented:
ping @vishvananda @eyakubovich

eyakubovich commented on link.go:
```diff
@@ -28,6 +28,14 @@ type LinkAttrs struct {
+		MTU:    -1,
+	}
+}
+
```
You're checking for `> 0` below so why not have "not set" be `== 0`. This way, default initialization still works and `NewLinkAttrs()` is not needed.

LK4D4 replied:
Hmm, my mistake. I need `>=0`. Is it true for MTU too?

LK4D4 replied:
No, I got it. For `MTU` 0 is sorta invalid.

eyakubovich replied:
`TxQLen` can be 0?

LK4D4 replied:
Yeah, at least I can set it with `iproute2`.

LK4D4 replied:
But that's weird, I can't set 0 actually with our lib :/

eyakubovich replied:
Looks like TxQLen == 0 has important special meaning in some drivers. Actually, same for MTU (basically says default to lower dev MTU). I guess we need `NewLinkAttrs`. @vishvananda thoughts?

LK4D4 replied:
Hmm, I'm not sure, but seems like we can't set `TxQLen` to `0` with netlink. `iproute2` uses `ioctl`.

vishvananda replied:
If we can't actually set zero for either then we could just leave zero as
the default, but I thought the initial bug was that it doesn't default to
10000

Vish

On Fri, Jun 12, 2015, 2:52 PM Alexander Morozov <notifications@github.com>
wrote:

> In link.go
> <https://github.com/vishvananda/netlink/pull/35#discussion_r32358495>:
>
> > @@ -28,6 +28,14 @@ type LinkAttrs struct {
> >  	Namespace    interface{} // nil | NsPid | NsFd
> >  }
> >
> > +// NewLinkAttrs returns LinkAttrs structure filled with default values
> > +func NewLinkAttrs() LinkAttrs {
> > +	return LinkAttrs{
> > +		TxQLen: -1,
> > +		MTU:    -1,
> > +	}
> > +}
> > +
>
> Hmm, I'm not sure, but seems like we can't set TxQLen to 0 with netlink.
> iproute2 uses ioctl.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/vishvananda/netlink/pull/35/files#r32358495>.
>

LK4D4 replied:
https://github.com/torvalds/linux/blob/master/net/core/rtnetlink.c#L1702

LK4D4 replied:
I was pretty sure, that we can :/ Actually with `TxQLen` all works as expected, apart from setting it to `0`.

eyakubovich replied:
But from that kernel code, it looks like 0 is ok and should be set. `tb[IFLA_TXQLEN` is just `rtattr *`

LK4D4 replied:
Yes, sorry, I'll retry my tests. TxQLen should be set.

eyakubovich commented:
+1 for checking for `> 0` before setting the attribute.

LK4D4 commented:
Okay, I think I found it. For veth pair, `TxQLen` is set to 1000 by default, but with netlink you're changing only peer `TxQLen`. For `dummy` type default is 0.

LK4D4 commented:
@eyakubovich @vishvananda Thanks for help!
Seems like it never worked! I changed code and added some tests.
MTU seems can't be zero, it returns `EINVAL` for Veth creation for example.

eyakubovich commented:
With veth, the main device and its peer are two separate devices (they each have ifindex). So you can't change TxQLen on them individually? Or did you mean that at creation time, TxQLen affects only peer.

LK4D4 commented:
No, I was wrong. Logic of txqlen was broken for main device. 
Also you can set different qlens for main and peer,  you can try with iproute2.

LK4D4 commented:
And original test for txqlen was false positive. I think all should be OK now.

vishvananda commented:
The code in the tests and the readme should be changed to use NewLinkAttrs. Unfortunately this means the nice one liners have to go away.

One option for allowing one liners would be to allow the New to take and return a pointer to a LinkAttrs that modifies the default:

http://play.golang.org/p/d4ZYrukfDE

but it has the drawback that NewLinkAttrs(&LinkAttrs{TxQLen: 0}) doesn't work as expected.

LK4D4 commented:
People still can use oneliners if they will set TxQLen explicitly. I'll write about this in Readme.

LK4D4 commented:
@vishvananda I added info to readme, but changing tests is a lot of work and not really worth, because we really don't care about `TxQLen` in most tests.
